[ML] Redesign file-based Data Visualizer

qn895 opened the pull request:
## Summary

This PR is a follow up of https://github.com/elastic/kibana/pull/85726 which updates the file based data visualizer to follow the new design. It also updates the functional tests for file based data visualizer.

**Before**
![Data Visualizer - Machine Learning - Elastic (8)](https://user-images.githubusercontent.com/43350163/103834343-bf29ff00-5048-11eb-9978-635563641e88.png)

**After**
![Data Visualizer - Machine Learning - Elastic (12)](https://user-images.githubusercontent.com/43350163/104658796-29662380-5689-11eb-973c-8c2fb7588277.png)

**Todo/Follow up PRs**
- [ ] Better support for geo fields (will do in a follow up PR)
### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/)
* Commit: fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/57afacfef221e020d1babe232d10d65c60e99902) | [after](https://github.com/elastic/kibana/commit/fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e) | diff |
| --- | --- | --- | --- |
| `ml` | 1690 | 1695 | +5 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/57afacfef221e020d1babe232d10d65c60e99902) | [after](https://github.com/elastic/kibana/commit/fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e) | diff |
| --- | --- | --- | --- |
| `ml` | 7.0MB | 7.0MB | -24.2KB |

#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/57afacfef221e020d1babe232d10d65c60e99902) | [after](https://github.com/elastic/kibana/commit/fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e) | diff |
| --- | --- | --- | --- |
| `default` | 47262 | 48022 | +760 |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/","number":"97462","commit":"fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e"}]}
PIPELINE-->

pheyos commented on x-pack/test/functional/apps/ml/data_visualizer/files_to_import/artificial_server_log:
```diff
@@ -1,19 +1,20 @@
-2018-01-06 17:36:15.295748 INFO host:'Server A' Connection from ip 123.456.789.1 closed
-2018-01-06 17:36:16.295748 INFO host:'Server A' Connection from ip 123.456.789.2 closed
-2018-01-06 17:36:17.295748 INFO host:'Server A' Connection from ip 123.456.789.3 closed
-2018-01-06 17:46:11.295748 INFO host:'Server B' Some special characters °!"§$%&/()=?`'^²³{[]}\+*~#'-_.:,;µ|<>äöüß
```
If possible, I'd like to keep some special characters in the log sample that we're loading, just to make sure it loads ok.

pheyos commented on x-pack/test/functional/services/ml/data_visualizer_table.ts:
```diff
@@ -320,5 +323,17 @@ export function MachineLearningDataVisualizerTableProvider(
+
+    public async ensureNumRowsPerPage(n: 10 | 25 | 100) {
+      await retry.tryForTime(10000, async () => {
+        await testSubjects.existOrFail('tablePaginationPopoverButton');
```
This is working fine at the moment, but will only work as long as there's only one table on the page, otherwise it might pick the wrong pagination control. I'd suggest to make this specific to the data visualizer table by nesting the test subject: `mlDataVisualizerTable > tablePaginationPopoverButton`.
Also, this is used three times in this method, so it could be a local const.

pheyos commented on x-pack/test/functional/services/ml/data_visualizer_table.ts:
```diff
@@ -320,5 +323,17 @@ export function MachineLearningDataVisualizerTableProvider(
+        await testSubjects.click(`tablePagination-${n}-rows`);
+
+        expect(await testSubjects.getVisibleText('tablePaginationPopoverButton')).to.contain(
+          n.toString()
```
This `contain` could potentially give a false positive here. E.g. if we're expecting `10` but the actual value is `100`, this validation will still succeed. Maybe it'd be better to take the full text, split on `:` and compare with `eql`?

pheyos reviewed (commented):
One of my test files ([ts_head.log](https://github.com/elastic/kibana/files/5781521/ts_head.log)) crashes the ML app on data visualizer load (blank page, have to switch to a different Kibana app and  come back to ML again). The console shows the following error:
![image](https://user-images.githubusercontent.com/1945390/103891633-7aef3b00-50ea-11eb-8d94-1c91581f4265.png)

peteharverson commented:
Feedback from initial testing:
- can the 'show distribution' histograms setting be retained after applying any overrides? When I apply overrides the distribution charts switch back to being hidden.

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/stats_datagrid.tsx:
```diff
@@ -20,60 +20,57 @@ import {
 }
 
-export const DataVisualizerDataGrid = ({
+export const DataVisualizerDataGrid = <T extends DataVisualizerDataGridItem>({
```
Nit - but I think this folder should be renamed `stats_data_grid`, or even just `data_grid` ?

jgowdyelastic replied:
the word "grid" isn't really relevant anymore as this PR turns it into a table.

qn895 replied:
Renamed the word DataGrid to `Table` here https://github.com/elastic/kibana/pull/87598/commits/ea57967e89c25c03a7de83aec9f802caf78c1708

jgowdyelastic replied:
the file names are still called "grid"

jgowdyelastic commented on x-pack/plugins/ml/public/application/datavisualizer/file_based/components/fields_stats_grid/fields_stats_grid.tsx:
```diff
@@ -0,0 +1,335 @@
+import { DataVisualizerFieldNamesFilter } from '../field_names_filter';
+import { DataVisualizerFieldTypesFilter } from '../field_types_filter';
+
+function createFields(results: FindFileStructureResponse) {
```
`createFields` and `filterFields ` are pretty large functions and could be moved out of this file to improve readability.

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/87598/commits/09f2b31a87c750f5c0ad60a849ee9e71242b923e

jgowdyelastic commented on x-pack/plugins/ml/public/application/datavisualizer/file_based/components/fields_stats_grid/get_field_names.ts:
```diff
@@ -29,3 +31,19 @@ export function getFieldNames(results) {
+
+export function getSupportedFieldType(type: string): MlJobFieldType {
+  switch (type) {
+    case 'long':
```
we can use `ES_FIELD_TYPES` from the data plugin here.

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/87598/commits/47dcf86095f2e0f9e3d78b6cb4c5ebbf5c17cd94

jgowdyelastic commented on x-pack/plugins/ml/common/types/file_datavisualizer.ts:
```diff
@@ -19,6 +19,60 @@ export interface AnalysisResult {
   overrides?: FormattedOverrides;
 }
 
+type NumericType =
```
The data plugin has these strings defined that we could import.
Annoyingly they defined them in their public location, and importing into common causes a linting error.
It seems a shame to redefine them.

jgowdyelastic replied:
I've just noticed that these types aren't used at all. I think they should be removed and if we do need them in the future we can base them on the ones exported by the data plugin.

jgowdyelastic commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_row/index.ts:
```diff
@@ -4,4 +4,4 @@
  */
 
-export { FieldsStats } from './fields_stats';
+export {};
```
was this left in by accident?

qn895 replied:
Removed here https://github.com/elastic/kibana/pull/87598/commits/8b99e8dd9f2085066c415ff58863d4aaec4b6f37

jgowdyelastic commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/ip_content.tsx:
```diff
@@ -9,15 +9,16 @@ import { EuiSpacer } from '@elastic/eui';
   if (stats === undefined) return null;
   const { count, sampleCount, cardinality } = stats;
   if (count === undefined || sampleCount === undefined || cardinality === undefined) return null;
+  const fieldFormat = 'fieldFormat' in config ? config.fieldFormat : undefined;
```
`const fieldFormat = config.fieldFormat` would be the same here.
Unless this ternary is to show that we're expecting `config.fieldFormat` to be undefined? Otherwise it seems redundant.

qn895 replied:
This is an annoying typescript quirk. This component is now used for both index based and file based data visualizer, and the file based props doesn't have `fieldFormat`. Unfortunately, even if subsequent components understand `fieldFormat` is optional, ts still give the error.
<img width="1043" alt="Screen Shot 2021-01-11 at 16 36 55" src="https://user-images.githubusercontent.com/43350163/104246275-51f0e200-542b-11eb-9980-38fc7ca8da3a.png">

mdefazio commented:
I haven't gone too far into this yet, but at first glance, will there be additional tabs alongside 'File stats' in the future? Otherwise I wonder if we just use the same heading styles as 'File content's and 'Summary' instead of the tab component.

jgowdyelastic commented:
> I haven't gone too far into this yet, but at first glance, will there be additional tabs alongside 'File stats' in the future? Otherwise I wonder if we just use the same heading styles as 'File content's and 'Summary' instead of the tab component.

If i remember correctly, the plan was to originally have some additional information here, but I can't see this happening.
I'd vote for switching it to a standard header.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/)
* Commit: 0a51062ef009eafedf580dc3f022000a4c1b0c73
* This comment will update when the build is complete

### Failed CI Steps
* [Check i18n](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/execution/node/381/log)

### History
* :green_heart: [Build #97462](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/) succeeded fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/","number":"97462","commit":"fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/","number":"98132","commit":"0a51062ef009eafedf580dc3f022000a4c1b0c73"}]}
PIPELINE-->

pheyos reviewed (approved):
Functional tests LGTM
Also, loading and importing of my example file is working again.

qn895 commented:
@peteharverson @mdefazio Thanks all for testing. I have made the changes which now include:
- Retain 'show distribution' histograms after applying any overrides
- Default show distribution to be on for file based visualizer
- Added some info in the collapsed row to indicate min / max / median for number type fields
- Added the true/false histogram for the expanded row and collapsed row for boolean type for both index and file base
<img width="1748" alt="Screen Shot 2021-01-12 at 17 55 22" src="https://user-images.githubusercontent.com/43350163/104388703-5d154200-54ff-11eb-958d-c7f913dc744b.png">

- Add histogram preview for IP types (both file and index)
<img width="1774" alt="Screen Shot 2021-01-12 at 17 56 29" src="https://user-images.githubusercontent.com/43350163/104388797-8e8e0d80-54ff-11eb-88f9-94e0fd639426.png">
<img width="1774" alt="Screen Shot 2021-01-12 at 17 57 29" src="https://user-images.githubusercontent.com/43350163/104388840-a796be80-54ff-11eb-8afd-a00d46e06c9f.png">

- Apply the stylistic changes from https://github.com/elastic/kibana/pull/87937
- Remove the tabs for the `File stats` since we decided it's no longer needed, and replace it with same heading as the other ones on the page
![Data Visualizer - Machine Learning - Elastic (11)](https://user-images.githubusercontent.com/43350163/104388664-3951fc00-54ff-11eb-953a-2c5ad03f37f0.png)

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/)
* Commit: 514f4a990514462c1e29463aab118dfb214ddacf
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/execution/node/363/log)

### History
* :broken_heart: [Build #98132](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/) failed 0a51062ef009eafedf580dc3f022000a4c1b0c73
* :green_heart: [Build #97462](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/) succeeded fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/","number":"97462","commit":"fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/","number":"98132","commit":"0a51062ef009eafedf580dc3f022000a4c1b0c73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/","number":"98426","commit":"514f4a990514462c1e29463aab118dfb214ddacf"}]}
PIPELINE-->

peteharverson commented:
Have you tried setting a max width for the components in the expanded rows? I wonder how that will look compared to the full width used for the charts for boolean and keyword types for example? 

For the boolean types, what about adding another column in the expanded row which gives the counts of `true` and `false`?

What are you thoughts @mdefazio?

![image](https://user-images.githubusercontent.com/7405507/104477910-93af9300-55b9-11eb-8598-5bf7ae3a61f6.png)

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/file_based/components/results_view/results_view.tsx:
```diff
@@ -103,7 +90,17 @@ export const ResultsView: FC<Props> = ({
+            </EuiTitle>
+
+            <FieldsStatsGrid results={results} />
+            {/* <EuiTabbedContent tabs={tabs} initialSelectedTab={tabs[0]} onTabClick={() => {}} />*/}
```
Is this commented out content needed?

qn895 replied:
Thanks for catching that. Removed here https://github.com/elastic/kibana/pull/87598/commits/bc29f197bd057f642315caaa2180a615712b9670#diff-5650b96d55c82f01621ecc4766b2e04e67fbc2083c567bb57cac92a8518eac68L103.

peteharverson reviewed (commented):
This is looking good. Just left a comment about whether a max / proportional width could be used on the expanded row charts.

mdefazio commented:
> Have you tried setting a max width for the components in the expanded rows? I wonder how that will look compared to the full width used for the charts for boolean and keyword types for example?
> For the boolean types, what about adding another column in the expanded row which gives the counts of true and false?

I think it could be a good idea—certainly on large screens the value on the chart is very far away from the label. Are we thinking a max pixel value or put in a max percentage for larger breakpoints?
Your second comment I think is worth doing as well—but maybe going further? We could also (re)display the type, document % value and distinct value as well to help add/condense the columns. So when you expand the row, it gives you a full summary, not just the charts / min max values.

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/file_based/components/fields_stats_grid/fields_stats_grid.tsx:
```diff
@@ -0,0 +1,202 @@
+          </EuiFlexGroup>
+        )}
+
+        <EuiFlexItem style={{ alignItems: 'flex-end' }}>
+          <EuiFlexGroup gutterSize="xs" data-test-subj="mlDataVisualizerFieldCountPanel">
```
I think you can use the `alignItems` prop in `EuiFlexGroup ` instead?

qn895 replied:
You're right. Updated here https://github.com/elastic/kibana/pull/87598/commits/a640650c72b5d0289aa0842622fa9b28fdcb1828.

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/file_based/components/fields_stats_grid/filter_fields.ts:
```diff
@@ -0,0 +1,40 @@
+export function filterFields(
+  fields: Array<
+    | FileBasedFieldVisConfig
+    | {
+        fieldName: string;
+        type: 'text' | 'unknown';
+        stats: { mean: number; count: number; sampleCount: number; cardinality: number };
+      }
```
worth extracting to an interface

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/87598/commits/93d64ccbcc8659622005d9447e3371942c3fef40

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/stats_table/data_visualizer_stats_table.tsx:
```diff
@@ -19,61 +19,62 @@ import {
+interface DataVisualizerTableProps<T> {
+  items: T[];
+  pageState: DataVisualizerIndexBasedAppState | DataVisualizerFileBasedAppState;
+  updatePageState: (update: any) => void;
```
please avoid using `any` if possible. `update` can be `Partial<DataVisualizerIndexBasedAppState  | DataVisualizerFileBasedAppState>` I presume

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/87598/commits/e6f1c5f39d01540e0a24875062722a36360390f1

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/stats_table/hooks/use_data_viz_chart_theme.ts:
```diff
@@ -0,0 +1,53 @@
+import type { PartialTheme } from '@elastic/charts';
+import { useUiSettings } from '../../../contexts/kibana';
+export const useDataVizChartTheme = (): PartialTheme => {
+  const IS_DARK_THEME = useUiSettings().get('theme:darkMode');
```
not sure if we need to follow this naming convention for such constants 
```suggestion
  const isDarkTheme = useUiSettings().get('theme:darkMode');
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/87598/commits/bccd4e7ecfdb4f789e60f9ca14d909af97da1b61

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/stats_table/types/field_vis_config.ts:
```diff
@@ -70,3 +70,22 @@ export interface FieldVisConfig {
+export function isIndexBasedFieldVisConfig(
+  field: FieldVisConfig | FileBasedFieldVisConfig
+): field is FieldVisConfig {
+  return (field as FieldVisConfig).existsInDocs !== undefined;
```
not sure if manual typecasting is a good approach in type guards

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/87598/commits/e6f1c5f39d01540e0a24875062722a36360390f1

qn895 commented on x-pack/plugins/ml/common/types/ml_url_generator.ts:
```diff
@@ -57,6 +57,17 @@ export interface DataVisualizerIndexBasedAppState {
   showEmptyFields?: boolean;
 }
+
+export interface DataVisualizerFileBasedAppState {
```
Updated here https://github.com/elastic/kibana/pull/87598/commits/18fad5d8c5809863d478ccd0c926c5e40b988128

qn895 commented on x-pack/plugins/ml/public/application/datavisualizer/file_based/components/fields_stats_grid/fields_stats_grid.tsx:
```diff
@@ -0,0 +1,202 @@
+interface Props {
+  results: FindFileStructureResponse;
+}
+export interface DataVisualizerFileBasedAppState {
```
Updated here https://github.com/elastic/kibana/pull/87598/commits/18fad5d8c5809863d478ccd0c926c5e40b988128

qn895 commented on x-pack/plugins/ml/public/application/datavisualizer/file_based/components/fields_stats_grid/fields_stats_grid.tsx:
```diff
@@ -0,0 +1,202 @@
+          <EuiFlexGroup
+            gutterSize="s"
+            alignItems="center"
+            style={{ maxWidth: 250 }}
```
Updated here https://github.com/elastic/kibana/pull/87598/commits/b70f474702004294d68a008d32bb114f81f9605a

qn895 commented on x-pack/plugins/ml/public/application/datavisualizer/file_based/components/field_data_row/number_content_preview.tsx:
```diff
@@ -0,0 +1,55 @@
+  )
+    return null;
+  return (
+    <EuiFlexGroup direction={'column'} gutterSize={'xs'}>
```
They should all be aligned by default by Eui

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/)
* Commit: a640650c72b5d0289aa0842622fa9b28fdcb1828
* This comment will update when the build is complete

### Failed CI Steps
* [Check i18n](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/execution/node/371/log)

### History
* :broken_heart: [Build #98426](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/) failed 514f4a990514462c1e29463aab118dfb214ddacf
* :broken_heart: [Build #98132](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/) failed 0a51062ef009eafedf580dc3f022000a4c1b0c73
* :green_heart: [Build #97462](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/) succeeded fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/","number":"97462","commit":"fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/","number":"98132","commit":"0a51062ef009eafedf580dc3f022000a4c1b0c73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/","number":"98426","commit":"514f4a990514462c1e29463aab118dfb214ddacf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/","number":"99017","commit":"a640650c72b5d0289aa0842622fa9b28fdcb1828"}]}
PIPELINE-->

mdefazio reviewed (changes requested):
I think this is quite close! Just a few comments: 

1. It looks like we have different gutter sizes set on the flex group for the expanded row depending on the field type.
Keyword:
![image](https://user-images.githubusercontent.com/3756330/104671214-5e836d80-56ab-11eb-9bb1-1a57ba362a0d.png)

Number:
![image](https://user-images.githubusercontent.com/3756330/104671240-693e0280-56ab-11eb-919f-bb70f757b379.png)

2. It might be nice to force the stats table to have a max-width and then the subsequent columns will start at the same point

![image](https://user-images.githubusercontent.com/3756330/104672624-1e71ba00-56ae-11eb-85b6-eeecdcb99695.png)

With a width set (instead of a percent value for the columns). 
![image](https://user-images.githubusercontent.com/3756330/104672789-7dcfca00-56ae-11eb-8ef3-f35fd5262540.png)

qn895 commented:
Thanks @mdefazio I have updated the gutter size to all match in the latest changes as well as having a fixed pixel number for the `Document stats` column
![Data Visualizer - Machine Learning - Elastic (13)](https://user-images.githubusercontent.com/43350163/104749983-11d67b80-5719-11eb-8989-b648efc5a01d.png)

peteharverson reviewed (approved):
Tested latest edits and LGTM

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99158/)
* Commit: e296ac090c384a01e05497336f5f894df53143b5

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/fc370b74eb2f5efc0fe58ef587371fc276d54bd6) | [after](https://github.com/elastic/kibana/commit/e296ac090c384a01e05497336f5f894df53143b5) | diff |
| --- | --- | --- | --- |
| `ml` | 1694 | 1712 | +18 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/fc370b74eb2f5efc0fe58ef587371fc276d54bd6) | [after](https://github.com/elastic/kibana/commit/e296ac090c384a01e05497336f5f894df53143b5) | diff |
| --- | --- | --- | --- |
| `ml` | 7.1MB | 7.0MB | -23.3KB |

### History
* :broken_heart: [Build #99017](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/) failed a640650c72b5d0289aa0842622fa9b28fdcb1828
* :broken_heart: [Build #98426](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/) failed 514f4a990514462c1e29463aab118dfb214ddacf
* :broken_heart: [Build #98132](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/) failed 0a51062ef009eafedf580dc3f022000a4c1b0c73
* :green_heart: [Build #97462](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/) succeeded fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/","number":"97462","commit":"fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/","number":"98132","commit":"0a51062ef009eafedf580dc3f022000a4c1b0c73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/","number":"98426","commit":"514f4a990514462c1e29463aab118dfb214ddacf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/","number":"99017","commit":"a640650c72b5d0289aa0842622fa9b28fdcb1828"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99158/","number":"99158","commit":"e296ac090c384a01e05497336f5f894df53143b5"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99248/)
* Commit: e69af6a5ae812a7ae5d72b989a1a1932f6316a79

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/5b38816a97b4f2bdd43700b46a94a5371cfc90ae) | [after](https://github.com/elastic/kibana/commit/e69af6a5ae812a7ae5d72b989a1a1932f6316a79) | diff |
| --- | --- | --- | --- |
| `ml` | 1694 | 1712 | +18 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/5b38816a97b4f2bdd43700b46a94a5371cfc90ae) | [after](https://github.com/elastic/kibana/commit/e69af6a5ae812a7ae5d72b989a1a1932f6316a79) | diff |
| --- | --- | --- | --- |
| `ml` | 7.1MB | 7.0MB | -22.9KB |

### History
* :green_heart: [Build #99158](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99158/) succeeded e296ac090c384a01e05497336f5f894df53143b5
* :broken_heart: [Build #99017](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/) failed a640650c72b5d0289aa0842622fa9b28fdcb1828
* :broken_heart: [Build #98426](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/) failed 514f4a990514462c1e29463aab118dfb214ddacf
* :broken_heart: [Build #98132](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/) failed 0a51062ef009eafedf580dc3f022000a4c1b0c73
* :green_heart: [Build #97462](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/) succeeded fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/97462/","number":"97462","commit":"fcf0d25ef0e47a8989a62c0e0fe8bc63bbc0221e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/","number":"98132","commit":"0a51062ef009eafedf580dc3f022000a4c1b0c73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/","number":"98426","commit":"514f4a990514462c1e29463aab118dfb214ddacf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/","number":"99017","commit":"a640650c72b5d0289aa0842622fa9b28fdcb1828"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99158/","number":"99158","commit":"e296ac090c384a01e05497336f5f894df53143b5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99248/","number":"99248","commit":"e69af6a5ae812a7ae5d72b989a1a1932f6316a79"}]}
PIPELINE-->

jgowdyelastic commented on x-pack/plugins/ml/public/application/datavisualizer/stats_table/hooks/use_data_viz_chart_theme.ts:
```diff
@@ -0,0 +1,53 @@
+import type { PartialTheme } from '@elastic/charts';
+import { useUiSettings } from '../../../contexts/kibana';
+export const useDataVizChartTheme = (): PartialTheme => {
+  const isDarkTheme = useUiSettings().get('theme:darkMode');
```
To be using `useUiSettings` in a hook like this, shouldn't it be watching for changes in the settings?
Like this function:
https://github.com/elastic/kibana/blob/f7fdda5db7e41b04f0f37cc180b441e8ed0c9321/x-pack/plugins/ml/public/application/components/color_range_legend/use_color_range.ts#L192

Otherwise it doesn't need to be in a hook.

qn895 replied:
You're right. This is now updated here https://github.com/elastic/kibana/pull/87598/commits/e09dc6ff939c54663d632f74d76e4f2caea4a9eb

jgowdyelastic reviewed (approved):
LGTM

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99613/)
* Commit: e09dc6ff939c54663d632f74d76e4f2caea4a9eb

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/7c9b11476dddbfdc27acefd27741744d3c9f202e) | [after](https://github.com/elastic/kibana/commit/e09dc6ff939c54663d632f74d76e4f2caea4a9eb) | diff |
| --- | --- | --- | --- |
| `ml` | 1694 | 1712 | +18 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/7c9b11476dddbfdc27acefd27741744d3c9f202e) | [after](https://github.com/elastic/kibana/commit/e09dc6ff939c54663d632f74d76e4f2caea4a9eb) | diff |
| --- | --- | --- | --- |
| `ml` | 7.1MB | 7.0MB | -23.1KB |

### History
* :green_heart: [Build #99248](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99248/) succeeded e69af6a5ae812a7ae5d72b989a1a1932f6316a79
* :green_heart: [Build #99158](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99158/) succeeded e296ac090c384a01e05497336f5f894df53143b5
* :broken_heart: [Build #99017](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/) failed a640650c72b5d0289aa0842622fa9b28fdcb1828
* :broken_heart: [Build #98426](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/) failed 514f4a990514462c1e29463aab118dfb214ddacf
* :broken_heart: [Build #98132](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/) failed 0a51062ef009eafedf580dc3f022000a4c1b0c73

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98132/","number":"98132","commit":"0a51062ef009eafedf580dc3f022000a4c1b0c73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/","number":"98426","commit":"514f4a990514462c1e29463aab118dfb214ddacf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/","number":"99017","commit":"a640650c72b5d0289aa0842622fa9b28fdcb1828"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99158/","number":"99158","commit":"e296ac090c384a01e05497336f5f894df53143b5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99248/","number":"99248","commit":"e69af6a5ae812a7ae5d72b989a1a1932f6316a79"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99613/","number":"99613","commit":"e09dc6ff939c54663d632f74d76e4f2caea4a9eb"}]}
PIPELINE-->

mdefazio reviewed (approved):
Tested locally and this all looks good.

darnautov reviewed (approved):
LGTM 🚀

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/100033/)
* Commit: f46544e67472d115e421f7daf51f88d78471b9c2

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/ad8589eb1911eb8918420194cefc85de8173b603) | [after](https://github.com/elastic/kibana/commit/f46544e67472d115e421f7daf51f88d78471b9c2) | diff |
| --- | --- | --- | --- |
| `ml` | 1694 | 1712 | +18 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/ad8589eb1911eb8918420194cefc85de8173b603) | [after](https://github.com/elastic/kibana/commit/f46544e67472d115e421f7daf51f88d78471b9c2) | diff |
| --- | --- | --- | --- |
| `ml` | 7.1MB | 7.0MB | -23.1KB |

### History
* :green_heart: [Build #99613](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99613/) succeeded e09dc6ff939c54663d632f74d76e4f2caea4a9eb
* :green_heart: [Build #99248](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99248/) succeeded e69af6a5ae812a7ae5d72b989a1a1932f6316a79
* :green_heart: [Build #99158](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99158/) succeeded e296ac090c384a01e05497336f5f894df53143b5
* :broken_heart: [Build #99017](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/) failed a640650c72b5d0289aa0842622fa9b28fdcb1828
* :broken_heart: [Build #98426](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/) failed 514f4a990514462c1e29463aab118dfb214ddacf

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/98426/","number":"98426","commit":"514f4a990514462c1e29463aab118dfb214ddacf"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99017/","number":"99017","commit":"a640650c72b5d0289aa0842622fa9b28fdcb1828"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99158/","number":"99158","commit":"e296ac090c384a01e05497336f5f894df53143b5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99248/","number":"99248","commit":"e69af6a5ae812a7ae5d72b989a1a1932f6316a79"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99613/","number":"99613","commit":"e09dc6ff939c54663d632f74d76e4f2caea4a9eb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/100033/","number":"100033","commit":"f46544e67472d115e421f7daf51f88d78471b9c2"}]}
PIPELINE-->
